### ISSUE TITLE
@aws-cdk/aws-cloudfront: Unable to deploy with EdgeFunction

### ISSUE THREAD

**Issue by asnaseer-resilient** [aws/aws-cdk#13644]
I can successfully deploy a CloudFront distribution using the CDK. However, as soon as I try to add a Lambda@Edge function to this, my deployment fails.

### Reproduction Steps
Here is the Typescript code I used to deploy:
```
import * as cdk from '@aws-cdk/core';
import * as cdkCertificateManager from '@aws-cdk/aws-certificatemanager';
import * as cdkCloudfront from '@aws-cdk/aws-cloudfront';
import * as cdkLambda from '@aws-cdk/aws-lambda';
import * as cdkOrigins from '@aws-cdk/aws-cloudfront-origins';
import * as cdkRoute53 from '@aws-cdk/aws-route53';
import * as cdkRoute53Targets from '@aws-cdk/aws-route53-targets';

import { CustomStackProps } from '@resilientplc/hubble-cdk-utils';

import { CDK_UTILS } from '../config/config';

import { CfProps } from './cf-props';

export class CfConstruct extends cdk.Construct {
  constructor(
    stack: cdk.Stack,
    {
      stage,
      env,
      tags,
      customProps: { hostedZoneId, hostedZoneDomainName, certificateArn, vpcName, graphQlEndpoint, apiR53RecordName, hubbleAlertsS3BucketName }
    }: CustomStackProps<CfProps>
  ) {
    super(stack, 'CloudFrontConstruct');

    const lambdaAtEdgeFunction = new cdkCloudfront.experimental.EdgeFunction(this, 'LambdaAtEdgeFunction', {
      runtime: cdkLambda.Runtime.NODEJS_12_X,
      code: cdkLambda.Code.fromAsset('.build/request-routing-lambda'),
      handler: 'request-routing-lambda.requestRoutingLambda'
    });

    const distribution = new cdkCloudfront.Distribution(this, 'CloudFrontDist', {
      comment: `${CDK_UTILS.resourceName('cloudfront')}: CloudFront distribution for GraphQL endpoint`,
      domainNames: [`${apiR53RecordName}.${hostedZoneDomainName}`],
      certificate: cdkCertificateManager.Certificate.fromCertificateArn(this, 'Certificate', certificateArn),
      minimumProtocolVersion: cdkCloudfront.SecurityPolicyProtocol.TLS_V1,
      priceClass: cdkCloudfront.PriceClass.PRICE_CLASS_100,
      defaultBehavior: {
        allowedMethods: cdkCloudfront.AllowedMethods.ALLOW_ALL,
        cachePolicy: new cdkCloudfront.CachePolicy(this, 'CloudFrontCachePolicy', {
          cachePolicyName: CDK_UTILS.resourceName('cloudfront-cache-policy'),
          headerBehavior: cdkCloudfront.CacheHeaderBehavior.allowList('Authorization'),
          defaultTtl: cdk.Duration.seconds(0),
          minTtl: cdk.Duration.seconds(0),
          maxTtl: cdk.Duration.seconds(1)
        }),
        edgeLambdas: [
          {
            eventType: cdkCloudfront.LambdaEdgeEventType.ORIGIN_REQUEST,
            functionVersion: lambdaAtEdgeFunction.currentVersion
          }
        ],
        origin: new cdkOrigins.HttpOrigin(cdk.Fn.select(2, cdk.Fn.split('/', graphQlEndpoint)), {
          originPath: '',
          originSslProtocols: [cdkCloudfront.OriginSslPolicy.TLS_V1, cdkCloudfront.OriginSslPolicy.TLS_V1_1, cdkCloudfront.OriginSslPolicy.TLS_V1_2]
        }),
        originRequestPolicy: new cdkCloudfront.OriginRequestPolicy(this, 'CloudFrontOriginRequestPolicy', {
          originRequestPolicyName: CDK_UTILS.resourceName('cloudfront-origin-request-policy'),
          headerBehavior: cdkCloudfront.OriginRequestHeaderBehavior.allowList('Origin', 'Access-Control-Request-Method', 'Access-Control-Request-Headers'),
          queryStringBehavior: cdkCloudfront.OriginRequestQueryStringBehavior.all(),
          cookieBehavior: cdkCloudfront.OriginRequestCookieBehavior.all()
        }),
        viewerProtocolPolicy: cdkCloudfront.ViewerProtocolPolicy.HTTPS_ONLY
      }
    });

    const zone = cdkRoute53.HostedZone.fromHostedZoneAttributes(this, 'R53HostedZone', { hostedZoneId, zoneName: hostedZoneDomainName });

    new cdkRoute53.ARecord(this, 'ARecord', {
      zone,
      recordName: apiR53RecordName,
      target: cdkRoute53.RecordTarget.fromAlias(new cdkRoute53Targets.CloudFrontTarget(distribution))
    });
  }
}
```

### What did you expect to happen?
I expected it to deploy without errors.

### What actually happened?
The deployment failed with the following errors:
```
cf (users-api-cf-dev): deploying...
[0%] start: Publishing 45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241:current
[100%] success: Published 45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241:current
users-api-cf-dev: creating CloudFormation changeset...
 0/7 | 13:11:30 | UPDATE_IN_PROGRESS   | AWS::CloudFormation::Stack               | users-api-cf-dev User Initiated
 1/7 | 13:11:57 | UPDATE_IN_PROGRESS   | AWS::CDK::Metadata                       | cf/CDKMetadata/Default (CDKMetadata) 
 1/7 | 13:11:57 | CREATE_IN_PROGRESS   | AWS::IAM::Role                           | cf/Custom::CrossRegionStringParameterReaderCustomResourceProvider/Role (CustomCrossRegionStringParameterReaderCustomResourceProviderRole71CD6825) 
 1/7 | 13:11:57 | CREATE_IN_PROGRESS   | AWS::IAM::Role                           | cf/Custom::CrossRegionStringParameterReaderCustomResourceProvider/Role (CustomCrossRegionStringParameterReaderCustomResourceProviderRole71CD6825) Resource creation Initiated
 1/7 | 13:11:58 | UPDATE_COMPLETE      | AWS::CDK::Metadata                       | cf/CDKMetadata/Default (CDKMetadata) 
 2/7 | 13:12:16 | CREATE_COMPLETE      | AWS::IAM::Role                           | cf/Custom::CrossRegionStringParameterReaderCustomResourceProvider/Role (CustomCrossRegionStringParameterReaderCustomResourceProviderRole71CD6825) 
 3/7 | 13:12:18 | CREATE_IN_PROGRESS   | AWS::Lambda::Function                    | cf/Custom::CrossRegionStringParameterReaderCustomResourceProvider/Handler (CustomCrossRegionStringParameterReaderCustomResourceProviderHandler65B5F33A) 
 3/7 | 13:12:19 | CREATE_IN_PROGRESS   | AWS::Lambda::Function                    | cf/Custom::CrossRegionStringParameterReaderCustomResourceProvider/Handler (CustomCrossRegionStringParameterReaderCustomResourceProviderHandler65B5F33A) Resource creation Initiated
 3/7 | 13:12:19 | CREATE_COMPLETE      | AWS::Lambda::Function                    | cf/Custom::CrossRegionStringParameterReaderCustomResourceProvider/Handler (CustomCrossRegionStringParameterReaderCustomResourceProviderHandler65B5F33A) 
 3/7 | 13:12:21 | CREATE_IN_PROGRESS   | Custom::CrossRegionStringParameterReader | cf/CloudFrontConstruct/LambdaAtEdgeFunction/ArnReader/Default (CloudFrontConstructLambdaAtEdgeFunctionArnReader5C1A0822) 
 3/7 | 13:12:27 | CREATE_IN_PROGRESS   | Custom::CrossRegionStringParameterReader | cf/CloudFrontConstruct/LambdaAtEdgeFunction/ArnReader/Default (CloudFrontConstructLambdaAtEdgeFunctionArnReader5C1A0822) Resource creation Initiated
 3/7 | 13:12:27 | CREATE_FAILED        | Custom::CrossRegionStringParameterReader | cf/CloudFrontConstruct/LambdaAtEdgeFunction/ArnReader/Default (CloudFrontConstructLambdaAtEdgeFunctionArnReader5C1A0822) Failed to create resource. ParameterNotFound: null
    at Request.extractError (/var/runtime/node_modules/aws-sdk/lib/protocol/json.js:52:27)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:688:14)
    at Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/var/runtime/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /var/runtime/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:690:12)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:116:18)
        new CustomResource (/Users/anjum/Documents/workspace/resilient/users-api/node_modules/@aws-cdk/core/lib/custom-resource.js:30:25)
        \_ EdgeFunction.createCrossRegionArnReader (/Users/anjum/Documents/workspace/resilient/users-api/node_modules/@aws-cdk/aws-cloudfront/lib/experimental/edge-function.js:255:26)
        \_ EdgeFunction.createCrossRegionFunction (/Users/anjum/Documents/workspace/resilient/users-api/node_modules/@aws-cdk/aws-cloudfront/lib/experimental/edge-function.js:232:30)
        \_ new EdgeFunction (/Users/anjum/Documents/workspace/resilient/users-api/node_modules/@aws-cdk/aws-cloudfront/lib/experimental/edge-function.js:43:20)
        \_ new CfConstruct (/Users/anjum/Documents/workspace/resilient/users-api/.build/stacks/cf/cf-construct.js:35:38)
        \_ new CfStack (/Users/anjum/Documents/workspace/resilient/users-api/.build/stacks/cf/cf-stack.js:28:9)
        \_ Object.<anonymous> (/Users/anjum/Documents/workspace/resilient/users-api/.build/stacks/stacks.js:52:1)
        \_ Module._compile (internal/modules/cjs/loader.js:774:30)
        \_ Object.Module._extensions..js (internal/modules/cjs/loader.js:785:10)
        \_ Module.load (internal/modules/cjs/loader.js:641:32)
        \_ Function.Module._load (internal/modules/cjs/loader.js:556:12)
        \_ Function.Module.runMain (internal/modules/cjs/loader.js:837:10)
        \_ internal/main/run_main_module.js:17:11
 3/7 | 13:12:28 | UPDATE_ROLLBACK_IN_P | AWS::CloudFormation::Stack               | users-api-cf-dev The following resource(s) failed to create: [CloudFrontConstructLambdaAtEdgeFunctionArnReader5C1A0822]. 
...
```

### Environment

  - **CDK CLI Version  : 1.91.0**
  - **Framework Version: 1.91.0**
  - **Node.js Version: v12.4.0**
  - **OS               : macOS Big Sur Version 11.2.3 (20D91)**
  - **Language (Version): Typescript (4.2.2)**

### Other
I have also raised this as a question in the slack channel: https://cdk-dev.slack.com/archives/C018XT6REKT/p1615987569068600

--- 

This is :bug: Bug Report

**Comment by ekeyser** [comment]
FYI I was able to deploy something with the following code. It's slightly modified from yours since I don't have your helper libs and config code. Also redacted some stuff. Deployed it twice. The TTL for cache policy is commented out because I think I've got a ver conflict with cdk-core. LMK if you think that's part of the issue and I can try to resolve (I don't think it is).

I'm running all same versions as you but I'm not on OSX. I can try there also if this doesn't help resolve the issue.

The error indicates (I think) the inability of your CloudFormation stack to read the ARN of your Edge function using the cross region reader. You can see some of this in the CloudFormation output yaml.

The below code deploys 2 stacks in different regions. Are you attempting the same or is everything in us-east-1?

```
import * as cdk from '@aws-cdk/core';
import * as cdkCertificateManager from '@aws-cdk/aws-certificatemanager';
import * as cdkCloudfront from '@aws-cdk/aws-cloudfront';
import * as cdkLambda from '@aws-cdk/aws-lambda';
import * as cdkOrigins from '@aws-cdk/aws-cloudfront-origins';
import * as cdkRoute53 from '@aws-cdk/aws-route53';
import * as cdkRoute53Targets from '@aws-cdk/aws-route53-targets';

export class CdkTests2Stack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const domainName = 'domain.tld';
    const uid: string = `${Date.now()}`;
    const FQDN: string = `${uid}.${domainName}`;
    const certificateArn = 'arn:aws:acm:us-east-1:*********:certificate/****************';
    const hostedZoneDomainName = domainName;
    const hostedZoneId = 'Z0560104************';
    const apiR53RecordName = 'github-testing';
    const graphQlEndpoint = 'https://blah.example.com/hey/happy';


    const lambdaAtEdgeFunction = new cdkCloudfront.experimental.EdgeFunction(this, 'LambdaAtEdgeFunction', {
      runtime: cdkLambda.Runtime.NODEJS_12_X,
      code: cdkLambda.Code.fromAsset('.build/request-routing-lambda'),
      // handler: 'request-routing-lambda.requestRoutingLambda'
      handler: 'index.handler'
    });

    const distribution = new cdkCloudfront.Distribution(this, 'CloudFrontDist', {
      comment: `stand-in: CloudFront distribution for GraphQL endpoint`,
      domainNames: [FQDN],
      certificate: cdkCertificateManager.Certificate.fromCertificateArn(this, 'Certificate', certificateArn),
      minimumProtocolVersion: cdkCloudfront.SecurityPolicyProtocol.TLS_V1,
      priceClass: cdkCloudfront.PriceClass.PRICE_CLASS_100,
      defaultBehavior: {
        allowedMethods: cdkCloudfront.AllowedMethods.ALLOW_ALL,
        cachePolicy: new cdkCloudfront.CachePolicy(this, 'CloudFrontCachePolicy', {
          cachePolicyName: `cache-policy-name`,
          headerBehavior: cdkCloudfront.CacheHeaderBehavior.allowList('Authorization'),
          // defaultTtl: Duration.seconds(300)
          // defaultTtl: cdk.Duration.seconds(30000000),
          // minTtl: cdk.Duration.seconds(0),
          // maxTtl: cdk.Duration.seconds(100)
        }),
        edgeLambdas: [
          {
            eventType: cdkCloudfront.LambdaEdgeEventType.ORIGIN_REQUEST,
            functionVersion: lambdaAtEdgeFunction.currentVersion
          }
        ],
        origin: new cdkOrigins.HttpOrigin(cdk.Fn.select(2, cdk.Fn.split('/', graphQlEndpoint)), {
          originPath: '',
          originSslProtocols: [cdkCloudfront.OriginSslPolicy.TLS_V1, cdkCloudfront.OriginSslPolicy.TLS_V1_1, cdkCloudfront.OriginSslPolicy.TLS_V1_2]
        }),
        originRequestPolicy: new cdkCloudfront.OriginRequestPolicy(this, 'CloudFrontOriginRequestPolicy', {
          originRequestPolicyName: 'cloudfront-origin-request-policy-name',
          headerBehavior: cdkCloudfront.OriginRequestHeaderBehavior.allowList('Origin', 'Access-Control-Request-Method', 'Access-Control-Request-Headers'),
          queryStringBehavior: cdkCloudfront.OriginRequestQueryStringBehavior.all(),
          cookieBehavior: cdkCloudfront.OriginRequestCookieBehavior.all()
        }),
        viewerProtocolPolicy: cdkCloudfront.ViewerProtocolPolicy.HTTPS_ONLY
      }
    });

    const zone = cdkRoute53.HostedZone.fromHostedZoneAttributes(this, 'R53HostedZone', {
      hostedZoneId,
      zoneName: hostedZoneDomainName
    });

    new cdkRoute53.ARecord(this, 'ARecord', {
      zone,
      recordName: apiR53RecordName,
      target: cdkRoute53.RecordTarget.fromAlias(new cdkRoute53Targets.CloudFrontTarget(distribution))
    });
  }
}
```

**Comment by ekeyser** [comment]
Wondering if you might want to upload the contents of the yaml from the failed deploy. Should still be accessible from the console. LMK if you need more guidance.

**Comment by asnaseer-resilient** [comment]
Here is a zip of the **cf.template.json** and **edge-lambda-stack-c89f6a7f843ae85632db3fd63885eb52a963e2d8ed.template.json** files that were produced as part of the deployment. I redacted the AWS account IDs in these files.

[Archive.zip](https://github.com/aws/aws-cdk/files/6159124/Archive.zip)

**Comment by asnaseer-resilient** [comment]
My stacks are deployed in the **eu-west-2** region (apart from the edge function which I believe should be automatically deployed into the **us-east-1** region).

**Comment by ekeyser** [comment]
Spent some time looking at this and nothing jumped out at me. Will take another look later.

**Comment by asnaseer-resilient** [comment]
@ekeyser Thank you for taking the time to look into this issue - greatly appreciated 👍

**Comment by asnaseer-resilient** [comment]
FYI: I tried to comment out the **defaultTtl**, **minTtl**, and **maxTtl** in the **cachePolicy** like you did but still get the same error.

**Comment by njlynch** [comment]
```
 3/7 | 13:12:27 | CREATE_FAILED        | Custom::CrossRegionStringParameterReader | cf/CloudFrontConstruct/LambdaAtEdgeFunction/ArnReader/Default (CloudFrontConstructLambdaAtEdgeFunctionArnReader5C1A0822) Failed to create resource. ParameterNotFound: null
```

The error (`ParameterNotFound: null`) makes me think that the SSM Parameter isn't being created properly in the us-east-1 stack. This deployment log is from your "main" (`eu-west-2`) stack. Can you check to see if the us-east-1 stack deployed successfully? You should have a SSM Parameter in us-east-1 called `EdgeFunctionArnLambdaAtEdgeFunction`; does that exist? Is it possible that the role you're using to deploy has some restrictions on it that might prevent it from gaining the `ssm:GetParameter` permission(s)?

**Comment by asnaseer-resilient** [comment]
@njlynch I do not see that parameter in the us-east-1 region.

**Comment by njlynch** [comment]
So either the deployment of the `edge-lambda-stack-*` stack failed, or it was never deployed. Do you have a successful deployment of the edge lambda stack in us-east-1? Is it possible that stack was never deployed? How did you deploy the app as a whole? (Generally, running `cdk deploy` in your situation will first deploy the us-east-1 stack, then your eu-west-2 stack).

**Comment by asnaseer-resilient** [comment]
This is the command I use for deployment of my stack named **cf**:
```
cdk deploy cf --app .build/stacks/stacks.js -o .build/cdk.out/cf --no-staging --exclusively --require-approval never --progress=events
```
And I do not see any evidence of the lambda@edge being deployed in the us-east-1 region in the AWS CloudFormation Console.

**Comment by asnaseer-resilient** [comment]
In case it helps, this is what I see in the CloudFormation Console in eu-west-2 when I try to deploy the above:
<img width="2014" alt="Screenshot 2021-03-18 at 13 39 35" src="https://user-images.githubusercontent.com/6410094/111635560-84360b80-87ef-11eb-9b86-a52b88985c6f.png">

**Comment by ekeyser** [comment]
Wasn't able to reproduce problem with OSX. Had correct ver w no conflicts (1.91.0) this time. Also put back the ttl params without issue.

Interesting how you are deploying. I'll have to play around to get things working like that. I usually just do a `cdk synth/deploy`. Is it correct for me to assume  these js files are after you execute an `npm run build` command? I take it `--exclusively` is to only deploy the specified stack and not any dependencies, correct?

I'm w @njlynch on this - it feels like the cross region reader resource is not able to read the edge function version (or some other function param perhaps) because the edge function doesn't exist (or if it were deployed separately maybe the expected resource names are misaligned or different).

**Comment by njlynch** [comment]
```
cdk deploy cf --app .build/stacks/stacks.js -o .build/cdk.out/cf --no-staging --exclusively ...
```

The `exclusively` flag is documented as:

```
  -e, --exclusively          Only deploy requested stacks, don't include
                             dependencies
```

This means you are *only* deploying the eu-west-2 stack, and not the us-east-1 stack. Without the us-east-1 stack, the Edge Function cannot work. Try deploying again, without the `--exclusively` flag.

**Comment by asnaseer-resilient** [comment]
I am going to try to create the edge lambda function in a separate stack which is deployed without the `--exclusively` flag. I will let you know if this works.

**Comment by njlynch** [comment]
> If I remove the --exclusively flag then will they all be deployed rather only the one we specify via the cf name directly after cdk deploy?

If you specify the stack, but remove the `--exclusively` flag, then the stack you specify, and all stacks which are necessary to deploy the stack you specified, will be deployed.

> We have some stacks in there that pass along their exported stack variables to other stacks. 

If Stack A exports a variable and Stack B consumes it, then Stack B is dependent on Stack A. If you deploy Stack B, Stack A will be automatically deployed (without the exclusively flag); if you deploy Stack A, Stack B will not be deployed.

**Comment by asnaseer-resilient** [comment]
Ok - everything now works as expected. Thank you all for your assistance with this 👍

**Comment by asnaseer-resilient** [comment]
Sorry to be a pain but I have hit a new related issue now. Here is a snippet of the top-level code in `stacks.ts`:
```
...

const lambdaEdgeStack = new LambdaEdgeStack(app, 'rrl', getStackProps('rrl', {}));

new CfStack(
  app,
  'cf',
  getStackProps('cf', {
    hostedZoneId: CONFIG.hostedZoneId,
    hostedZoneDomainName: CONFIG.hostedZoneDomainName,
    certificateArn: CONFIG.certificateArn,
    graphQlEndpoint: appSyncLegacyStack.exportedStackVariables.graphQlEndpoint.value as UrlString,
    apiR53RecordName: CONFIG.apiR53RecordName,
    lambdaEdgeVersion: lambdaEdgeStack.version
  })
);
```
I managed to deploy the Lambda@Edge function stack using:
`cdk deploy rrl --app .build/stacks/stacks.js -o .build/cdk.out/rrl --no-staging --require-approval never --progress=events`

and then deployed the CloudFront distribution stack using this:
`cdk deploy cf --app .build/stacks/stacks.js -o .build/cdk.out/cf --no-staging --exclusively --require-approval never --progress=events`

However, when I then made a small change to my Lambda@Edge implementation code and tried to deploy it again I got:
```
deploying rrl to dev
Including dependency stacks: users-api-lambda-edge-dev
users-api-lambda-edge-dev
users-api-lambda-edge-dev: deploying...
[0%] start: Publishing 89159adf48d27038819243c0a94aa985bd6e564ffa852f73722c6eb86d4fc35a:current
[50%] success: Published 89159adf48d27038819243c0a94aa985bd6e564ffa852f73722c6eb86d4fc35a:current
[50%] start: Publishing 67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24:current
[100%] success: Published 67b7823b74bc135986aa72f889d6a8da058d0c4a20cbc2dfc6f78995fdd2fc24:current
users-api-lambda-edge-dev: creating CloudFormation changeset...
 0/5 | 11:59:42 | UPDATE_IN_PROGRESS   | AWS::CloudFormation::Stack | users-api-lambda-edge-dev User Initiated
 1/5 | 12:00:10 | UPDATE_IN_PROGRESS   | AWS::Lambda::Function | LambdaAtEdgeFunction (LambdaAtEdgeFunction29DB2087) 
 1/5 | 12:00:11 | UPDATE_COMPLETE      | AWS::Lambda::Function | LambdaAtEdgeFunction (LambdaAtEdgeFunction29DB2087) 
 1/5 | 12:00:13 | CREATE_IN_PROGRESS   | AWS::Lambda::Version  | LambdaAtEdgeFunction/CurrentVersion (LambdaAtEdgeFunctionCurrentVersionE45A41DDd624ed08b82f550267823476b69a9834) 
 2/5 | 12:00:14 | CREATE_IN_PROGRESS   | AWS::Lambda::Version  | LambdaAtEdgeFunction/CurrentVersion (LambdaAtEdgeFunctionCurrentVersionE45A41DDd624ed08b82f550267823476b69a9834) Resource creation Initiated
 2/5 | 12:00:15 | CREATE_COMPLETE      | AWS::Lambda::Version  | LambdaAtEdgeFunction/CurrentVersion (LambdaAtEdgeFunctionCurrentVersionE45A41DDd624ed08b82f550267823476b69a9834) 
 2/5 | 12:00:17 | UPDATE_IN_PROGRESS   | AWS::SSM::Parameter   | LambdaAtEdgeFunction/Parameter (LambdaAtEdgeFunctionParameter8C21B1A8) 
 5/5 | 12:00:19 | UPDATE_COMPLETE      | AWS::SSM::Parameter   | LambdaAtEdgeFunction/Parameter (LambdaAtEdgeFunctionParameter8C21B1A8) 
 5/5 | 12:00:21 | UPDATE_COMPLETE_CLEA | AWS::CloudFormation::Stack | users-api-lambda-edge-dev 
 5/5 | 12:00:22 | DELETE_IN_PROGRESS   | AWS::Lambda::Version  | LambdaAtEdgeFunctionCurrentVersionE45A41DD924c47671a50c8d8090a48a17ba60496 
 5/5 | 12:00:23 | DELETE_COMPLETE      | AWS::Lambda::Version  | LambdaAtEdgeFunctionCurrentVersionE45A41DD924c47671a50c8d8090a48a17ba60496 
 5/5 | 12:00:23 | UPDATE_COMPLETE      | AWS::CloudFormation::Stack | users-api-lambda-edge-dev 

 ✅  users-api-lambda-edge-dev

Stack ARN:
arn:aws:cloudformation:us-east-1:*****:stack/users-api-lambda-edge-dev/e19a5800-8897-11eb-94e0-120e6de30a9d
rrl (users-api-rrl-dev)
rrl (users-api-rrl-dev): deploying...
[0%] start: Publishing 45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241:current
[100%] success: Published 45b7ed524ce2b119dd4f2b8642ae8bfaf0df45bc6bd705072ae4ee6d1a999241:current
users-api-rrl-dev: creating CloudFormation changeset...
 0/2 | 12:00:30 | UPDATE_IN_PROGRESS   | AWS::CloudFormation::Stack               | users-api-rrl-dev User Initiated
 0/2 | 12:01:00 | UPDATE_IN_PROGRESS   | Custom::CrossRegionStringParameterReader | rrl/LambdaEdgeConstruct/LambdaAtEdgeFunction/ArnReader/Default (LambdaEdgeConstructLambdaAtEdgeFunctionArnReader5988D08D) 
 1/2 | 12:01:07 | UPDATE_COMPLETE      | Custom::CrossRegionStringParameterReader | rrl/LambdaEdgeConstruct/LambdaAtEdgeFunction/ArnReader/Default (LambdaEdgeConstructLambdaAtEdgeFunctionArnReader5988D08D) 
 1/2 | 12:01:08 | UPDATE_ROLLBACK_IN_P | AWS::CloudFormation::Stack               | users-api-rrl-dev Export users-api-rrl-dev:ExportsOutputFnGetAttLambdaEdgeConstructLambdaAtEdgeFunctionArnReader5988D08DFunctionArn23D9794B cannot be updated as it is in use by users-api-cf-dev
 0/2 | 12:01:13 | UPDATE_IN_PROGRESS   | Custom::CrossRegionStringParameterReader | rrl/LambdaEdgeConstruct/LambdaAtEdgeFunction/ArnReader/Default (LambdaEdgeConstructLambdaAtEdgeFunctionArnReader5988D08D) 
 0/2 | 12:01:15 | UPDATE_COMPLETE      | Custom::CrossRegionStringParameterReader | rrl/LambdaEdgeConstruct/LambdaAtEdgeFunction/ArnReader/Default (LambdaEdgeConstructLambdaAtEdgeFunctionArnReader5988D08D) 
 1/2 | 12:01:16 | UPDATE_ROLLBACK_COMP | AWS::CloudFormation::Stack               | users-api-rrl-dev 
 1/2 | 12:01:16 | UPDATE_ROLLBACK_COMP | AWS::CloudFormation::Stack               | users-api-rrl-dev 

 ❌  rrl (users-api-rrl-dev) failed: Error: The stack named users-api-rrl-dev failed to deploy: UPDATE_ROLLBACK_COMPLETE
    at Object.waitForStackDeploy (/Users/anjum/Documents/workspace/resilient/users-api/node_modules/aws-cdk/lib/api/util/cloudformation.ts:307:11)
    at processTicksAndRejections (internal/process/task_queues.js:89:5)
    at Object.deployStack (/Users/anjum/Documents/workspace/resilient/users-api/node_modules/aws-cdk/lib/api/deploy-stack.ts:291:26)
    at CdkToolkit.deploy (/Users/anjum/Documents/workspace/resilient/users-api/node_modules/aws-cdk/lib/cdk-toolkit.ts:180:24)
    at initCommandLine (/Users/anjum/Documents/workspace/resilient/users-api/node_modules/aws-cdk/bin/cdk.ts:207:9)
The stack named users-api-rrl-dev failed to deploy: UPDATE_ROLLBACK_COMPLETE
```

Is there something I can do to fix this?

**Comment by asnaseer-resilient** [comment]
The `LambdaEdgeStack` constructs a `LambdaEdgeConstruct` which looks like this:
```
import * as cdk from '@aws-cdk/core';
import * as cdkCloudfront from '@aws-cdk/aws-cloudfront';
import * as cdkIam from '@aws-cdk/aws-iam';
import * as cdkLambda from '@aws-cdk/aws-lambda';

import { CustomStackProps } from '@resilientplc/hubble-cdk-utils';

import { CDK_UTILS } from '../config/config';

import { LambdaEdgeProps } from './lambda-edge-props';

export class LambdaEdgeConstruct extends cdk.Construct {
  public version: cdkLambda.IVersion;

  constructor(stack: cdk.Stack, { stage, env, tags }: CustomStackProps<LambdaEdgeProps>) {
    super(stack, 'LambdaEdgeConstruct');

    const lambdaAtEdgeFunction = new cdkCloudfront.experimental.EdgeFunction(this, 'LambdaAtEdgeFunction', {
      runtime: cdkLambda.Runtime.NODEJS_12_X,
      code: cdkLambda.Code.fromAsset('.build/request-routing-lambda'),
      functionName: CDK_UTILS.resourceName(env!.region!),
      stackId: CDK_UTILS.resourceName('lambda-edge'),
      description: `${tags!.product}-${stage} Lambda@Edge function for AppSync GraphQL requests`,
      handler: 'request-routing-lambda.requestRoutingLambda',
      initialPolicy: [
        new cdkIam.PolicyStatement({
          effect: cdkIam.Effect.ALLOW,
          actions: ['servicediscovery:DiscoverInstances'],
          resources: ['*']
        }),
        CDK_UTILS.lambdaCloudwatchLogsPolicyStatement()
      ],
      logRetention: 60,
      timeout: cdk.Duration.seconds(5),
      tracing: cdkLambda.Tracing.ACTIVE
    });

    this.version = lambdaAtEdgeFunction.currentVersion;
  }
}
```

**Comment by njlynch** [comment]
```
 1/2 | 12:01:08 | UPDATE_ROLLBACK_IN_P | AWS::CloudFormation::Stack               | users-api-rrl-dev Export users-api-rrl-dev:ExportsOutputFnGetAttLambdaEdgeConstructLambdaAtEdgeFunctionArnReader5988D08DFunctionArn23D9794B cannot be updated as it is in use by users-api-cf-dev
```

It looks like you have an export that is used by another dependency stack.

There are a few resources for how to handle this:
* The general CloudFormation approach - https://aws.amazon.com/premiumsupport/knowledge-center/cloudformation-stack-export-name-error/
* Using `stack.exportValue()` - https://github.com/aws/aws-cdk/pull/12778
* Another discussion on the same topic - https://github.com/aws/aws-cdk/issues/5304

I'm going to assign this to one of our fantastic front-line support guys who can help you with any follow-ups.

**Comment by asnaseer-resilient** [comment]
I merged the Lambda@Edge function back into the same stack that was creating the CloudFront distribution and it all works just fine now with no export issues.

One thing I have noticed however is that the edge function created by `new cdkCloudfront.experimental.EdgeFunction(...)` does not seem to have any of the tags that were setup in the stack properties. All other resources created via the CDK seem to have them. The edge function only seems to have these tags:

- **aws:cloudformation:stack-name**
- **aws:cloudformation:stack-id**
- **aws:cloudformation:logical-id**

Is this a known issue?
